torch/optim/adamw.py

```diff
@@ -101,14 +120,15 @@ def __setstate__(self, state):
             group.setdefault('foreach', None)
             group.setdefault('capturable', False)
             group.setdefault('differentiable', False)
+            group.setdefault('fused', False)
         state_values = list(self.state.values())
         step_is_tensor = (len(state_values) != 0) and torch.is_tensor(state_values[0]['step'])
         if not step_is_tensor:
             for s in state_values:
                 s['step'] = torch.tensor(float(s['step']))
 
     @_use_grad_for_differentiable
-    def step(self, closure=None):
+    def step(self, closure=None, *, grad_scaler=None):
```

Comment:
Hmmm do we have a larger plan for grad_scaler support?
In particular;
- Why was this not needed before?
- Will we want this to be added to all optimizer?
- Is this specific to cuda.amp or is it general for all amp?
- How is the user expected to use it and how is it different from what they do today?

Reply:
> * Why was this not needed before?

Because `_single_tensor_adam(w)` and `_multi_tensor_adam(w)` IIUC do not have an efficient means to fuse gradient unscaling (it's still feasible but it'd not be efficient than `torch.cuda.amp.GradScaler`'s unscaling which utilizes multi_tensor_apply basically).

> * Will we want this to be added to all optimizer?

As optimizers support fused path from my perspective.

> * Is this specific to cuda.amp or is it general for all amp?

This is specific to float16 amp requiring `torch.cuda.amp.GradScaler`.
IIRC CPU autocast only support BFloat16. Is this correct?

> * How is the user expected to use it and how is it different from what they do today?

I don't think there're user-facing changes in order to pass `grad_scaler` to this `step` method. 
Rather `GradScaler.step(optimizer)` checks a private attribute of an optimizer and pass itself as keyword argument: https://github.com/pytorch/pytorch/blob/049a0f2cd5916c8392c6bd1adc41c709de892f3a/torch/cuda/amp/grad_scaler.py#L327-L334

Reply:
> Rather GradScaler.step(optimizer) checks a private attribute of an optimizer and pass itself as keyword argument:

Ho wow that is super shady...
I think there are a few issues with this approach:
- It makes it very hard for users to write a custom optimizer "properly"
- It is quite brittle: what if a user changes the fused flag, the grad scaler will just be silently ignored.
- It makes the Optimizer and GradScaler code very much linked to each other. At this point why are they even different things?
- While the step() API can make sense, the updated API for the functional function is not ok. We should not use private objects there/have arguments that don't apply and are just returning a plain assert.

I think there are two directions forward for this:
- Refactor this in terms of generic concepts from the point of view of the Optimizer. Like we could have a gradient scale input as well as a mask for what should be updated or not. The GradScaler would then only update these values before calling the regular optimizer.
- Or we say that GradScaler cannot be separated from the Optimizer at all and need to be special cased. If so, then we should define clearly what the step function should do with these and then make it a first class citizen.

Side note, we are adding in https://github.com/pytorch/pytorch/pull/89176 hooks for the step function. So the GradScaler could update some of the params (like scale and mask) without having to override the optimizer and simply by registering a hook there.

Reply:
I'm not quite sure if I understand your points clearly

This check of `_step_supports_amp_scaling` done by `GradScaler.step(optimizer)` allows for the fusion of gradient unscaling and parameter update (i.e. reduce the number of CUDA kernels and synchronization) which then enables the combination of `GradScaler` and CUDA Graphs, and potentially low precision parameters (if gradient unscaling for fp16 tensors is done in a separate cuda kernel, the gradient scaling or loss scaling get meaningless).
This `fused` path is an example of "ninja optimizers" in "Interaction with New Custom Optimizers" of https://github.com/pytorch/pytorch/issues/25081

> It makes it very hard for users to write a custom optimizer "properly"

could you elaborate on "properly"?
If one doesn't set the attribute, then the behavior would be equivalent to the vanilla optimizers and multi_tensor optimizers.

> It is quite brittle: what if a user changes the fused flag, the grad scaler will just be silently ignored.

assuming that one instantiated AdamW with `fused=True` and change the `fused` after a certain number of `grad_scaler.step(optimizer)` calls, then it'd just error out because `torch.optim.adam.adam` will be `_single_tensor_adam`.

> It makes the Optimizer and GradScaler code very much linked to each other. At this point why are they even different things?

I think this design of `GradScaler` (and `Optimizer`) is the (or one of the) conclusion from https://github.com/pytorch/pytorch/issues/25081

Reply:
Thanks for sharing the original issue. That helps a lot to give context here!
I definitely agree that having such fusion is useful and important.
My question above was only how do we want to implement it.

> could you elaborate on "properly"?

I mean in such a way that it will work correctly and be fast. And when I say users, that includes us where we have one more thing to keep in mind for every change we do to our Optimizers.

> I think this design of GradScaler (and Optimizer) is the (or one of the) conclusion from

The design yes, but it doesn't seem that the implementation is defined there.
In particular, the two proposal I made above are both following the proposal from that issue just implementing it in a different way (to avoid magic attributes).

Reply:
> > could you elaborate on "properly"?
> 
> I mean in such a way that it will work correctly and be fast. And when I say users, that includes us where we have one more thing to keep in mind for every change we do to our Optimizers.

Thanks. I misunderstood what you meant by users.

---

> > I think this design of GradScaler (and Optimizer) is the (or one of the) conclusion from
> 
> The design yes, but it doesn't seem that the implementation is defined there. In particular, the two proposal I made above are both following the proposal from that issue just implementing it in a different way (to avoid magic attributes).

I think I completely agree with you in that you don't think it's okay to update `F.adam(w)` with weird type objects which I came up to satisfy TorchScript and let the behavior of optimizer.step and GradScaler depend on a hidden attribute of `"_step_supports_amp_scaling"`, but I'm yet to be certain what your suggestions (in the first comment) and the other comments indicate. So please bear with my questions, admitting I'm overfitted to this style of implementing as you can see in `torch.optim.Adam` and [apex's fused lamb](https://github.com/NVIDIA/apex/blob/3c8f5161a63b877b52d0161e09da367cdd8c5460/apex/optimizers/fused_mixed_precision_lamb.py).

- Would you want to clear off the dependency on `"_step_supports_amp_scaling"` and this attribute completely?
    - Without this look up of attribute, with the ongoing optimizer hooks (I'd assume the pre), and e.g. `optimizer.scale` and `optimizer.found_inf`, we could update these two tensors and loosen the link between GradScaler and Optimizer (still need to think about a way to let the scaler know it doesn't need to sync before calling optimizer.step)
- Would you be happy if `F.adamw` gets two new arguments of `grad_scale: Optional[Tensor]` and `found_inf[Tensor]`? It's still specific to CUDA FP16 AMP though I came up with this when I was rewriting the logic of grouping tensors by device & dtype in C++ (in case you're interested in, here it is: https://github.com/crcrpar/pytorch/blob/33c45c4b20c85a2294026ccbd442e62c3ef31959/torch/optim/adam.py#L254-L255)

Reply:
what's your take on the already updated Adam optimizer?

Reply:
It should be updated to follow the same semantic.